lib/common/ipc_server.c

```diff
@@ -757,16 +757,69 @@ pcmk__ipc_send_xml(pcmk__client_t *c, uint32_t request, const xmlNode *message,
 
     iov_buffer = g_string_sized_new(1024);
     pcmk__xml_string(message, 0, iov_buffer, 0);
-    rc = pcmk__ipc_prepare_iov(request, iov_buffer, 0, &iov, NULL);
 
-    if (rc == pcmk_rc_ok) {
-        pcmk__set_ipc_flags(flags, "send data", crm_ipc_server_free);
-        rc = pcmk__ipc_send_iov(c, iov, flags);
+    /* Testing crm_ipc_server_event is obvious.  pcmk__client_proxied is less
+     * obvious.  According to pcmk__ipc_send_iov, replies to proxied connections
+     * need to be sent as events.  However, do_local_notify (which calls this
+     * function) will clear all flags so we can't go just by crm_ipc_server_event.
+     *
+     * Changing do_local_notify to check for a proxied connection first results
+     * in processes on the Pacemaker Remote node (like cibadmin or crm_mon)
+     * timing out when waiting for a reply.
+     */
+    if (pcmk_is_set(flags, crm_ipc_server_event)
+        || pcmk_is_set(c->flags, pcmk__client_proxied)) {
+        uint16_t index = 0;
+
+        do {
+            rc = pcmk__ipc_prepare_iov(request, iov_buffer, index, &iov, NULL);
+
+            switch (rc) {
+                case pcmk_rc_ok:
+                    /* No more chunks to send after this one */
+                    pcmk__set_ipc_flags(flags, "send data", crm_ipc_server_free);
+                    rc = pcmk__ipc_send_iov(c, iov, flags);
```

Comment:
Here, if we get `EAGAIN`, we will return it. Do we want to map `EAGAIN` to `pcmk_rc_ok` before returning? We could let the callers handle it appropriately, but (at a quick glance), I don't think they do currently.

Reply:
I wish I hadn't looked at the callers - they're a mess.  Some pass FALSE for flags, some don't even handle the return code at all.  Still, I think it's worth doing that mapping.  That way the callers don't have to think about whether they passed an event or not when interpreting the return code.